### PR TITLE
MAINT: Use PyBIDS 0.9.x via niworkflows PR

### LINKED ISSUE #93: MAINT: Bump pybids to 0.9.x
To avoid conflicts when upgrading fMRIPrep.

### PR DESCRIPTION
**Pull request by effigies**
Closes #93.

### CODE REVIEW

**Comment by oesteban** [comment]
I'd be :+1: to move out to `setup.cfg` instead of ``requirements.txt`` + ``__about__.py``.

**Comment by effigies** [comment]
Probably best left to another PR. This one already has the potential to be fiddly.

**Comment by effigies** [comment]
I'm guessing the failures relate to the license not being available to PRs?

**Comment by oesteban** [comment]
I believe this error building the Docker image might be the original cause for ds054 to fail:

```
ERROR: templateflow 0.4.0 has requirement pybids<0.10.0a0,>=0.9.0, but you'll have pybids 0.2.1+1231.g0c39d71 which is incompatible.
Installing collected packages: indexed-gzip, nibabel, nilearn, pillow, imageio, decorator, networkx, PyWavelets, scikit-image, isodate, rdflib, lxml, prov, click, neurdflib, simplejson, pydotplus, funcsigs, future, packaging, pydot, nipype, docopt, num2words, patsy, bids-validator, SQLAlchemy, pybids, MarkupSafe, jinja2, tqdm, templateflow, PyYAML, seaborn, svgutils, niworkflows
Successfully installed MarkupSafe-1.1.1 PyWavelets-1.0.3 PyYAML-5.1.1 SQLAlchemy-1.3.5 bids-validator-1.2.4 click-7.0 decorator-4.4.0 docopt-0.6.2 funcsigs-1.0.2 future-0.17.1 imageio-2.5.0 indexed-gzip-0.8.10 isodate-0.6.0 jinja2-2.10.1 lxml-4.3.4 networkx-2.3 neurdflib-5.0.0.post1 nibabel-2.4.1 nilearn-0.5.2 nipype-1.2.0 niworkflows-0.9.0+78.gdc80ea8 num2words-0.5.10 packaging-19.0 patsy-0.5.1 pillow-6.1.0 prov-1.5.3 pybids-0.2.1+1231.g0c39d71 pydot-1.4.1 pydotplus-2.0.2 rdflib-4.2.2 scikit-image-0.15.0 seaborn-0.9.0 simplejson-3.16.0 svgutils-0.3.1 templateflow-0.4.0 tqdm-4.32.2
 ---> e828ddb6f9b9
Step 28/39 : RUN python -c "from templateflow import api as tfapi;                tfapi.get('MNI152Lin|MNI152NLin2009cAsym|OASIS30ANTs', suffix='T1w');                tfapi.get('MNI152Lin|MNI152NLin2009cAsym|OASIS30ANTs', desc='brain', suffix='mask');                tfapi.get('OASIS30ANTs', resolution=1, desc='4', suffix='dseg');                tfapi.get('OASIS30ANTs|NKI', resolution=1, label='brain', suffix='probseg');                tfapi.get('MNI152NLin2009cAsym|OASIS30ANTs|NKI', resolution=1, desc='BrainCerebellumRegistration', suffix='mask'); "
 ---> Running in e66ef92dc691
Downloading https://templateflow.s3.amazonaws.com/tpl-OASIS30ANTs/tpl-OASIS30ANTs_res-01_desc-4_dseg.nii.gz
```

**Comment by effigies** [comment]
My guess is it's the report template. That looks fine, actually, though I've pushed the tags to my branch to avoid it in the future.

**Comment by oesteban** [comment]
Not sure:

```
Node: smriprep_wf.single_subject_100185_wf.anat_preproc_wf.anat_norm_wf.tpl_mask
Working directory: /tmp/ds054/work/smriprep_wf/single_subject_100185_wf/anat_preproc_wf/anat_norm_wf/_template_MNI152Lin/tpl_mask

Node inputs:

args = <undefined>
default_value = 0.0
dimension = 3
environ = {'NSLOTS': '1'}
float = True
input_image = /tmp/ds054/work/smriprep_wf/single_subject_100185_wf/anat_preproc_wf/brain_extraction_wf/atropos_wf/copy_xform/09_relabel_wm_mask_xform.nii.gz
input_image_type = <undefined>
interpolation = MultiLabel
interpolation_parameters = <undefined>
invert_transform_flags = <undefined>
num_threads = 1
out_postfix = _trans
output_image = <undefined>
print_out_composite_warp_file = <undefined>
reference_image = /home/smriprep/.cache/templateflow/tpl-MNI152Lin/tpl-MNI152Lin_res-01_T1w.nii.gz
transforms = ['/tmp/ds054/work/smriprep_wf/single_subject_100185_wf/anat_preproc_wf/anat_norm_wf/_template_MNI152Lin/registration/ants_t1_to_mniComposite.h5']

Traceback (most recent call last):
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/plugins/multiproc.py", line 69, in run_node
    result['result'] = node.run(updatehash=updatehash)
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/engine/nodes.py", line 472, in run
    result = self._run_interface(execute=True)
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/engine/nodes.py", line 563, in _run_interface
    return self._run_command(execute)
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/pipeline/engine/nodes.py", line 643, in _run_command
    result = self._interface.run(cwd=outdir)
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/interfaces/base/core.py", line 375, in run
    runtime = self._run_interface(runtime)
  File "/usr/local/miniconda/lib/python3.7/site-packages/niworkflows/interfaces/fixes.py", line 23, in _run_interface
    runtime, correct_return_codes)
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/interfaces/base/core.py", line 758, in _run_interface
    self.raise_exception(runtime)
  File "/usr/local/miniconda/lib/python3.7/site-packages/nipype/interfaces/base/core.py", line 695, in raise_exception
    ).format(**runtime.dictcopy()))
RuntimeError: Command:
antsApplyTransforms --default-value 0 --dimensionality 3 --float 1 --input /tmp/ds054/work/smriprep_wf/single_subject_100185_wf/anat_preproc_wf/brain_extraction_wf/atropos_wf/copy_xform/09_relabel_wm_mask_xform.nii.gz --interpolation MultiLabel --output 09_relabel_wm_mask_xform_trans.nii.gz --reference-image /home/smriprep/.cache/templateflow/tpl-MNI152Lin/tpl-MNI152Lin_res-01_T1w.nii.gz --transform /tmp/ds054/work/smriprep_wf/single_subject_100185_wf/anat_preproc_wf/anat_norm_wf/_template_MNI152Lin/registration/ants_t1_to_mniComposite.h5
Standard output:

Standard error:
Exception caught during reference file reading 

itk::ExceptionObject (0x3afeb70)
Location: "unknown" 
File: /tmp/ANTs/build/ITKv4/Modules/IO/PNG/src/itkPNGImageIO.cxx
Line: 107
Description: itk::ERROR: PNGImageIO(0x3af58f0): PNGImageIO failed to read header for file: 
Reason: fread read only 0 instead of 8

 file /home/smriprep/.cache/templateflow/tpl-MNI152Lin/tpl-MNI152Lin_res-01_T1w.nii.gz
Segmentation fault (core dumped)
Return code: 139
```
Looks like `/home/smriprep/.cache/templateflow/tpl-MNI152Lin/tpl-MNI152Lin_res-01_T1w.nii.gz` is zero-size (which is how TemplateFlow from S3 works: there are these zero-size files and they get pulled down and replaced when requested via `get` for first).

**Comment by effigies** [comment]
Ah, you're right. I think that's just a general problem, in that templateflow doesn't always get files.

**Comment by effigies** [comment]
Restarted this one, in case it's the weird templateflow issue exhibiting in another way.

**Comment by effigies** [comment]
It was not. SSHing.

File: Dockerfile [hunk 1]
```diff
@@ -146,21 +147,20 @@ RUN useradd -m -s /bin/bash -G users smriprep
...
 RUN pip install --no-cache-dir -r requirements.txt && \
     rm -rf $HOME/.cache/pip
 
+# Precaching atlases
+RUN python -c "from templateflow import api as tfapi; \
+               tfapi.get(['MNI152Lin', 'MNI152NLin2009cAsym', 'OASIS30ANTs'], suffix='T1w'); \
+               tfapi.get(['MNI152Lin', 'MNI152NLin2009cAsym', 'OASIS30ANTs'], desc='brain', suffix='mask'); \
+               tfapi.get('OASIS30ANTs', resolution=1, desc='4', suffix='dseg'); \
+               tfapi.get(['OASIS30ANTs', 'NKI'], resolution=1, label='brain', suffix='probseg'); \
+               tfapi.get(['MNI152NLin2009cAsym', 'OASIS30ANTs', 'NKI'], resolution=1, desc='BrainCerebellumRegistration', suffix='mask'); "
```
Comment (effigies) [inline Dockerfile]: The surface issue appears to have been that the atlases were specified as regexes, not lists. The deeper issue is that templateflow is not being used to fetch these files at need.

**Comment by effigies** [comment]
The issue now is that `linked_svg = layout.build_path(entities)` [here](https://github.com/poldracklab/niworkflows/blob/master/niworkflows/reports/core.py#L155) is failing, producing `None`.

```
Traceback (most recent call last):
  File "/usr/local/miniconda/bin/smriprep", line 10, in <module>
    sys.exit(main())
  File "/usr/local/miniconda/lib/python3.7/site-packages/smriprep/cli/run.py", line 14, in main
    return build_opts(opts)
  File "/usr/local/miniconda/lib/python3.7/site-packages/smriprep/cli/run.py", line 265, in build_opts
    packagename='smriprep')
  File "/usr/local/miniconda/lib/python3.7/site-packages/niworkflows/reports/core.py", line 422, in generate_reports
    for subject_label in subject_list
  File "/usr/local/miniconda/lib/python3.7/site-packages/niworkflows/reports/core.py", line 422, in <listcomp>
    for subject_label in subject_list
  File "/usr/local/miniconda/lib/python3.7/site-packages/niworkflows/reports/core.py", line 409, in run_reports
    subject_id=subject_label, packagename=packagename)
  File "/usr/local/miniconda/lib/python3.7/site-packages/niworkflows/reports/core.py", line 258, in __init__
    self._load_config(Path(config))
  File "/usr/local/miniconda/lib/python3.7/site-packages/niworkflows/reports/core.py", line 276, in _load_config
    self.index(settings['sections'])
  File "/usr/local/miniconda/lib/python3.7/site-packages/niworkflows/reports/core.py", line 298, in index
    for cfg in subrep_cfg['reportlets']]
  File "/usr/local/miniconda/lib/python3.7/site-packages/niworkflows/reports/core.py", line 298, in <listcomp>
    for cfg in subrep_cfg['reportlets']]
  File "/usr/local/miniconda/lib/python3.7/site-packages/niworkflows/reports/core.py", line 158, in __init__
    out_file = out_dir / linked_svg
  File "/usr/local/miniconda/lib/python3.7/pathlib.py", line 892, in __truediv__
    return self._make_child((key,))
  File "/usr/local/miniconda/lib/python3.7/pathlib.py", line 679, in _make_child
    drv, root, parts = self._parse_args(args)
  File "/usr/local/miniconda/lib/python3.7/pathlib.py", line 633, in _parse_args
    a = os.fspath(a)
TypeError: expected str, bytes or os.PathLike object, not NoneType
```

If the cause is obvious to you, let me know. Otherwise I'll dig in in the morning.

**Comment by effigies** [comment]
`ValueError: Could not generate SVG path to copy /tmp/ds054/work/reportlets/smriprep/sub-100185/anat/sub-100185_space-MNI152Lin_t1w.svg to. Entities: {'datatype': 'figures', 'space': 'MNI152Lin', 'subject': '100185', 'suffix': 't1w', 'extension': 'svg'}`

Looks like it's the lowercase suffix causing the problem.

File: smriprep/workflows/norm.py [hunk 2]
```diff
@@ -239,7 +239,7 @@ def init_anat_norm_wf(
     norm_rpt.inputs.after_label = 'Participant'  # after
 
     ds_t1_2_tpl_report = pe.Node(
-        DerivativesDataSink(base_directory=reportlets_dir, keep_dtype=True),
+        DerivativesDataSink(base_directory=reportlets_dir, suffix='T1w'),
```
Comment (effigies) [inline smriprep/workflows/norm.py]: @oesteban If you understand why we were getting effective inputs as listed [here](https://619-160736148-gh.circle-artifacts.com/0/tmp/ds054/work/smriprep_wf/single_subject_100185_wf/anat_preproc_wf/anat_norm_wf_MNI152Lin/ds_t1_2_tpl_report/_report/report.rst), let me know. I can't figure out where `suffix` is getting defined as `'t1w'`, so I'm just overriding it.

Comment (oesteban) [reply smriprep/workflows/norm.py]: weird. I'm okay overriding it. We can come back to this later.

Comment (effigies) [reply smriprep/workflows/norm.py]: Unresolved. I think `DerivativesDataSink` must be overriding inputs, but it's really unclear where.

https://625-160736148-gh.circle-artifacts.com/0/tmp/ds054/work/smriprep_wf/single_subject_100185_wf/anat_preproc_wf/anat_norm_wf_MNI152Lin/ds_t1_2_tpl_report/_report/report.rst

Comment (oesteban) [reply smriprep/workflows/norm.py]: weird

Comment (oesteban) [reply smriprep/workflows/norm.py]: Why this is not happening for ds005? - maybe there're some cached results?

Comment (effigies) [reply smriprep/workflows/norm.py]: Are some of the working directories/derivatives directories cached?

Comment (oesteban) [reply smriprep/workflows/norm.py]: Yes - https://github.com/poldracklab/smriprep/blob/master/.circleci/config.yml#L428

We should re-run all the reportlet datasinks - i.e., remove the reportlets subfolder for both ds005 and ds054.

**Comment by effigies** [comment]
Okay, it was a cache issue. That's good. I'll go ahead and test fMRIPrep, and we should hopefully be done with all of this.

**Comment by oesteban** [comment]
Should we merge altogether with #98? Otherwise, is this still WIP?

**Comment by effigies** [comment]
Oh, nope, this is good to go. We can merge altogether or in two shots. The other one might be easier to review separately.

**Comment by oesteban** [comment]
Let's do it in two shots so we don't miss this one in the CHANGES log.

**Pull request merged**